Interactive bots: Create Maths bot by JefftheBest1.

Jeffthebest1 opened the pull request:
I have created a MathsBot which takes @math and @maths. It is easy to add operations to it

smarx commented:
*Automated message from Dropbox CLA bot*

@Jeffthebest1, it looks like you've already signed the Dropbox CLA. Thanks!

TomaszKolek commented:
Could you try to run `./tools/test-all` on your own? I'm almost sure that lint test will raise some errors.

TomaszKolek commented on contrib_bots/lib/maths.py:
```diff
@@ -0,0 +1,148 @@
+        charnum = 0 #Number of char in calculation
+        for i in calc:  #Look at each character
+            charnum += 1
+            if negnum == True:
```
`if negnum` is enough

TomaszKolek commented on contrib_bots/lib/maths.py:
```diff
@@ -0,0 +1,148 @@
+            if negnum == True:
+                i = "-"+i
+                negnum = False
+            if (i in operators) and (prevval is not None) and (prevval != "op") and charnum != len(calc):  #Check for valid operators
```
parenthesis are not needed here

TomaszKolek commented on contrib_bots/lib/maths.py:
```diff
@@ -0,0 +1,148 @@
+    
+    def do_maths(self,maths_nums,maths_ops):
+
+        maths_nums = list(maths_nums)
```
What is the case we need to convert it to `list`? It seems its always a list.

TomaszKolek commented on contrib_bots/lib/maths.py:
```diff
@@ -0,0 +1,148 @@
+        #If adding operators, remember to add them here ^^^^ as well
+        #Assign operators to the operator functions 
+        op_dict = {'*':operator.mul, '/':operator.div, '+':operator.add, '-':operator.sub} # You can add your own operators here
+        Value = None
```
variables should be lowercase always

TomaszKolek commented:
Did you try to find external library that parses math expressions? It might be more reliable.

Jeffthebest1 commented:
Ok, I've taken on board all your feedback. I didn't try to find an external library because I didn't think that I was allowed to. Anyway, I will do that now

Jeffthebest1 commented:
Wouldn't it be better just to keep the code I have?
 It works and it is easy to add new maths operations

TomaszKolek commented:
Please squash your commits - here should be just one commit.

Jeffthebest1 commented:
It's passed all the tests - it just needs Travis now

showell commented:
@Jeffthebest1 Can you give us a status update on this PR?  My understanding from our Zulip conversation from a couple days ago is that you're going to spend some time re-organizing the code to make it more testable.  Also, you and Joshua were having some discussion about implementing the expression evaluator differently--I'm not sure what came of that conversation.

If you are not super intent on getting these changed merge to zulip, then you can close this PR, and we can figure out a way to get your task approved.

Jeffthebest1 commented:
@showell I've added separate classes for the calculator and handler, added a testing function which runs automatically when the file is run as a library and I have just generally cleaned everything up. I have nearly finished an implementation of parentheses. Joshua decided not to help me.

showell commented:
Ok, just to be clear, you are still working on cleaning up the code and haven't re-pushed yet, correct?

Jeffthebest1 commented:
Correct

Jeffthebest1 commented:
I've now made the changes (the ones I listed above) and pushed them.

showell commented on contrib_bots/lib/maths.py:
```diff
@@ -0,0 +1,243 @@
+                buffer = []
+                brac_lvl -= 1
+            elif self.is_number(i):
+            # Stuff that isn't operators.These characters go to the buffer.
```
please indent this comment

showell commented:
I discussed this with @Jeffthebest1 on Zulip.  We are not going to merge this, because we probably eventually want to use something like Wolfram for this.   Closing.
